src/exo/cursors.py

```diff
@@ -0,0 +1,591 @@
+        """
+        assert self._path
+        assert len(self) > 0
+
+        def update(node):
+            attr, i = self._path[-1]
+            children = getattr(node, attr)
+            new_children = children[:i.start] + nodes + children[i.stop:]
+            new_children = new_children or empty_default or []
+            return node.update(**{attr: new_children})
+
+        from .API import Procedure
+        p = Procedure(self._rewrite_node(update))
+
+        return p, self._forward_replace(weakref.ref(p), len(nodes))
+
+    def _forward_replace(self, new_proc, n_ins):
+        del_range = self._path[-1][1]
+        n_diff = n_ins - len(del_range)
+
+        def fwd_node(i):
+            if i in del_range:
+                raise InvalidCursorError('node no longer exists')
+            return i + n_diff * (i >= del_range.stop)
+
+        def fwd_gap(i):
+            if i in del_range[1:]:
+                raise InvalidCursorError('gap no longer exists')
+            return i + n_diff * (i >= del_range.stop)
+
+        def fwd_sel(rng: range):
+            if _is_sub_range(rng, del_range):
+                raise InvalidCursorError('block no longer exists')
+            if _overlaps_one_side(rng, del_range):
+                raise InvalidCursorError('block was partially destroyed')
+
+            rng = range(
+                del_range.start if rng.start in del_range else rng.start,
+                del_range.stop if rng.stop in del_range[1:] else rng.stop
+            )
+
+            start = rng.start + n_diff * (rng.start >= del_range.stop)
+            stop = rng.stop + n_diff * (rng.stop >= del_range.stop)
+
+            if start >= stop:
+                raise InvalidCursorError('block no longer exists')
+
+            return range(start, stop)
+
+        return self._make_forward(new_proc, fwd_node, fwd_gap, fwd_sel)
+
+    def _delete(self):
+        """
+        This is an UNSAFE internal function for deleting a block in an AST and
+        providing a forwarding function as collateral. It is meant to be
+        package-private, not class-private, so it may be called from other
+        internal classes and modules, but not from end-user code.
+        """
+        pass_stmt = [LoopIR.Pass(None, self.parent()._node().srcinfo)]
+        return self._replace([], empty_default=pass_stmt)
+
+
+@dataclass
+class Node(Cursor):
+    _path: list[tuple[str, Optional[int]]]
+
+    # ------------------------------------------------------------------------ #
+    # Validating accessors
+    # ------------------------------------------------------------------------ #
+
+    def _node(self):
+        """
+        Gets the raw underlying node that's pointed-to. This is meant to be
+        compiler-internal, not class-private, so other parts of the compiler
+        may call this, while users should not.
+        """
+        if (n := self._node_ref()) is None:
+            raise InvalidCursorError('underlying node was destroyed')
```

Comment:
Should this be an internal assertion error?  If you intend for failed forwarding functions to raise `InvalidCursorError` eagerly during forwarding, then the only way you could trip this case is if the implementation of some forwarding function was broken.

Reply:
This seems even more pressing if I think about this method being called all over the `LoopIR_scheduling.py` code.  By the time we're doing things there, we probably oughtn't be triggering any more Cursor Invalidations

Reply:
As of right now, it's unreachable because the provenance system keeps everything alive. I think this will be worth thinking about more once that's fixed and this becomes reachable again.

Reply:
It's impossible _for now_. Eventually we will want users to know that they've attempted to use a cursor for a procedure that no longer exists... right?

Reply:
Like, a user gets a cursor into a proc for scheduling it, makes a bunch of changes, then calls `.partial_eval` to make a bunch of variants and lets the original proc die... cursors don't keep procs alive, so it would die here.